infrastructure/src/main/java/org/corfudb/infrastructure/LayoutServer.java

```diff
@@ -384,18 +402,18 @@ public void setCurrentLayout(Layout layout) {
         setLayoutInHistory(layout);
     }
 
-    public Rank getPhase1Rank() {
+    public Rank getPhase1Rank(long epoch) {
         return paxosDataStore
-                .getPhase1Rank(getServerEpoch())
+                .getPhase1Rank(epoch)
                 .orElse(null);
     }
 
-    public void setPhase1Rank(Rank rank) {
-        paxosDataStore.setPhase1Rank(rank, getServerEpoch());
+    public void setPhase1Rank(Rank rank, long epoch) {
+        paxosDataStore.setPhase1Rank(rank, epoch);
     }
 
-    public void setPhase2Data(Phase2Data phase2Data) {
-        paxosDataStore.setPhase2Data(phase2Data, getServerEpoch());
+    public void setPhase2Data(Phase2Data phase2Data, long epoch) {
+        paxosDataStore.setPhase2Data(phase2Data, epoch);
```

Comment:
I think its worthwhile to add some sanity checks here, for example `assert phase2Data.getLayout().getEpoch() == epoch` . Actually, if we had this check this the problem wouldn't have happened.

Reply:
Whether you do `assert phase2Data.getLayout().getEpoch() == epoch`or `if (msgEpoch != serverEpoch) return NACK` does not really matter since both of them are doing the same thing -- ensuring that epoch does not change between the time we do the validation and and the time we persist the accepted data.

Ideally all Paxos phases should base their decision on immutable state. So enforcing immutability at the beginning of each round seems cleaner to me.

Reply:
Its not about immutability on the server side. If there's a race condition on the client side -- where the message is being stamped with an epoch that is different than the layout's epoch (inside the message), then pre-check doesn't matter (i.e. it won't be detected).

Reply:
If you look at `LayoutView`, you will notice that within `updateLayout(Layout layout, long rank)` we do:

``` java
        long epoch = layout.getEpoch();
        ...
        //phase 1: prepare with a given rank.
        Layout alreadyProposedLayout = prepare(epoch, rank);
        ...
        //phase 2: propose the new layout.
        propose(epoch, rank, layoutToPropose);
        //phase 3: committed
        committed(epoch, layoutToPropose);
```

Which means that you can carry either:

1) Your layout to phase II. We know that these cannot be different since we extract the epoch directly from the layout: `long epoch = layout.getEpoch()`.
2) Already accepted layout to phase II. This one also cannot happen since we make sure to return the layout associated with the provided epoch, which cannot be different because of the previous proposition.

``` java
            // Return the layout with the highest rank proposed before.
            Rank highestProposedRank = proposedLayout == null ? new Rank(-1L, msg.getClientID())
                    : getPhase2Rank(msgEpoch);
```

Which means there is no way for the client for send us a layout with a different epoch. However, I do agree that we should probably check for that invariant as well. It will shield us  from future regressions.

Reply:
I see where the confusion is coming from. There are three (!) epoch in that message:

1) The layout epoch.
2) The payload epoch.
3) The message epoch.

We are not relying on the message epoch. (1) and (2) should probably be collapsed into one, but that is future work.